File: golang/datastructure/binarysearchtree.go
package datastructure

import (
	"fmt"
)

type binarySearchTreeNode struct {
	data interface{}
	score int64
	left *binarySearchTreeNode
	right *binarySearchTreeNode
}

type binarySearchTree struct {
	root *binarySearchTreeNode
	size int
}

func BinarySearchTree() *binarySearchTree {
	return &binarySearchTree{root:nil, size:0}
}

func (b *binarySearchTree) Add(data interface{}, score int64) {
	b.root = b.add(b.root, data, score)
}

func (b *binarySearchTree) add(node *binarySearchTreeNode, e interface{}, score int64) *binarySearchTreeNode {
	if (node == nil) {
		b.size++
		return &binarySearchTreeNode{data:e, score:score, left: nil, right: nil}
	}
	if score < node.score {
		node.left = b.add(node.left, e, score)
	}
	if score > node.score {
		node.right = b.add(node.right, e, score)
	}
	return node
}

// 前序遍历
func (b *binarySearchTree) PreOrder() {
	b.preOrder(b.root)
}

func (b *binarySearchTree) preOrder(node *binarySearchTreeNode) {
	if node == nil {
		return
	}
	fmt.Println(node.data)
	b.preOrder(node.left)
	b.preOrder(node.right)
}

// 百递归前序遍历
func (b *binarySearchTree) PreOrderNR() {
	b.preOrderNR(b.root)
}

func (b *binarySearchTree) preOrderNR(node *binarySearchTreeNode) {
	if b.root == nil {
		return
	}
	s := Stack()
	s.Push(b.root)

	for s.Size() > 0 {
		cur, _ := s.Pop().(*binarySearchTreeNode)
		fmt.Println(cur.data)

		if cur.left != nil {
			s.Push(cur.left)
		}
		if cur.right != nil {
			s.Push(cur.right)
		}
	}
}

// 中序遍历
func (b *binarySearchTree) InOrder() {
	b.inOrder(b.root)
}

func (b *binarySearchTree) inOrder(node *binarySearchTreeNode) {
	if node == nil {
		return
	}
	b.inOrder(node.left)
	fmt.Println(node.data)
	b.inOrder(node.right)
}

// 后序遍历
func (b *binarySearchTree) PostOrder() {
	b.postOrder(b.root)
}

func (b *binarySearchTree) postOrder(node *binarySearchTreeNode) {
	if node == nil {
		return
	}
	b.postOrder(node.left)
	b.postOrder(node.right)
	fmt.Println(node.data)
}

// 层序遍历
func (b *binarySearchTree) LevelOrder() {
	if b.root == nil {
		return
	}
	q := Queue()
	q.Push(*b.root)
	for q.Size() > 0 {
		cur, _ := q.Pop().(binarySearchTreeNode)
		fmt.Println(cur.data)

		if cur.left != nil {
			q.Push(*cur.left)
		}
		if cur.right != nil {
			q.Push(*cur.right)
		}
	}
}

// 获取最小元素
func (b *binarySearchTree) Min() interface{} {
	return b.min(b.root).data
}

func (b *binarySearchTree) min(node *binarySearchTreeNode) *binarySearchTreeNode {
	if node.left == nil {
		return node
	}
	return b.min(node.left)
}

// 获取最大值
func (b *binarySearchTree) Max() interface{} {
	return b.max(b.root)
}

func (b *binarySearchTree) max(node *binarySearchTreeNode) *binarySearchTreeNode {
	if node.right == nil {
		return node
	}
	return b.max(node.right)
}

// 删除最小元素
func (b *binarySearchTree) DelMin() interface{} {
	min := b.Min()
	b.root = b.delMin(b.root)
	return min
}

func (b *binarySearchTree) delMin(node *binarySearchTreeNode) *binarySearchTreeNode {
	if node.left == nil {
		tmpNode := node.right
		node.right = nil
		b.size--
		return tmpNode
	}
	node.left = b.delMin(node.left)
	return node
}

// 删除最大元素
func (b *binarySearchTree) DelMax() interface{} {
	max := b.Max()
	b.root = b.delMax(b.root)
	return max
}

func (b *binarySearchTree) delMax(node *binarySearchTreeNode) *binarySearchTreeNode {
	if node.right == nil {
		tmp := node.left
		node.right = nil
		b.size--
		return tmp
	}
	node.right = b.delMax(node.right)
	return node
}

// 删除任意元素（通过score）
func (b *binarySearchTree) Del(score int64) {
	b.root = b.del(b.root, score)
}

func (b *binarySearchTree) del(node *binarySearchTreeNode, score int64) *binarySearchTreeNode {
	if node == nil {
		return nil
	}
	if node.score > score {
		node.left = b.del(node.left, score)
		return node
	}
	if node.score < score {
		node.right = b.del(node.right, score)
		return node
	}
	// 找到元素
	if node.left == nil {
		tmp := node.right
		node.right = nil
		b.size--
		return tmp
	}
	if node.right == nil {
		tmp := node.left
		node.left = nil
		b.size--
		return tmp
	}

	// 到这里说明左右子树都不为nil
	// 将右子树最小元素放到当前要删除的元素位置
	minLeftNode := b.min(node.right)
	minLeftNode.right = b.delMin(node.right)
	minLeftNode.left = node.left

	node.left = nil
	node.right = nil

	return minLeftNode
}

File: java/BinarySearchTree/src/BinarySearchTree.java
import java.util.Stack;
import java.util.LinkedList;
import java.util.Queue;
public class BinarySearchTree<E extends Comparable<E>> {
    private class Node {
        public E e;
        public Node left,right;
        public Node(E e) {
            this.e = e;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;
    private int size;

    public BinarySearchTree() {
        this.root = null;
        this.size = 0;
    }

    public int size() {
        return size;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public void add(E e) {
        root = this.add(root, e);
    }

    private Node add(Node node, E e) {
        if (node == null) {
            size++;
            return new Node(e);
        }

        if (e.compareTo(node.e) < 0)
            node.left = this.add(node.left, e);
        else if (e.compareTo(node.e) > 0)
            node.right = this.add(node.right, e);
        return node;
    }

    public boolean contains(E e) {
        return contains(root, e);
    }

    private boolean contains(Node node, E e) {
        if (node == null)
            return false;

        if (e.compareTo(node.e) == 0) {
            return true;
        }  else if (e.compareTo(node.e) < 0) {
            return contains(node.left, e);
        } else {
            return contains(node.right, e);
        }
    }

    // 前序遍历
    public void preOrder() {
        preOrder(root);
    } 

    private void preOrder(Node node) {
        if (node == null)
            return;

        System.out.println(node.e);
        preOrder(node.left);
        preOrder(node.right);
    }

    // 非递归前序遍历
    public void preOrderNR() {
        if (root == null)
            return;
        
        Stack<Node> stack = new Stack<>();
        stack.push(root);
        while(!stack.isEmpty()) {
            Node cur = stack.pop();
            System.out.println(cur.e);
            if (cur.right != null)
                stack.push(cur.right);
            if (cur.left != null)
                stack.push(cur.left);
        }
    }

    // 中序遍历
    public void inOrder() {
        inOrder(root);
    }

    private void inOrder(Node node) {
        if (node == null) 
            return;
        inOrder(node.left);
        System.out.println(node.e);
        inOrder(node.right);
    }

    // 后序遍历 
    public void postOrder() {
        postOrder(root);
    }

    private void postOrder(Node node) {
        if (node == null)
            return;
        postOrder(node.left);
        postOrder(node.right);
        System.out.println(node.e);
    }

    // 层序遍历 
    public void levelOrder() {
        if (root == null)
            return;
        Queue<Node> q = new LinkedList<>();
        q.add(root);
        while(!q.isEmpty()) {
            Node cur = q.remove();
            
            System.out.println(cur.e);
            if (cur.left != null)
                q.add(cur.left);
            if (cur.right != null) 
                q.add(cur.right);
        }
    }

    // 获取最小值
    public E min() {
        return min(root).e;
    }

    private Node min(Node node) {
        if (node.left == null) 
            return node;

        return min(node.left);
    }

    // 获取最大值
    public E max() {
        return max(root).e;
    }

    private Node max(Node node) {
        if (node.left == null) 
            return node;
        return max(node.right);
    }

    // 删除最小值 
    public E delMin() {
        E e = min();
        root = delMin(root);
        return e;
    }

    // 删除最小值
    private Node delMin(Node node) {
        if (node.left == null) {
            Node tmpNode = node.right;
            node.right = null;
            size--;
            return tmpNode;
        }

        node.left = delMin(node.left);
        return node;
    }

    // 删除最大值
    public E delMax() {
        E e = max();
        root = delMax(root);
        return e;
    }

    private Node delMax(Node node) {
        if (node.right == null) {
            Node tmpNode = node.left;
            node.left = null;
            size--;
            return tmpNode;
        }
        node.right = delMax(node.right);
        return node;
    }

    // 删除任意节点
    public void del(E e) {
        root = del(root, e);   
    }

    private Node del(Node node, E e) {
        if (node == null) 
            return null;
        if (e.compareTo(node.e) < 0) {
            node.left = del(node.left, e);
            return node;
        } else if (e.compareTo(node.e) > 0) {
            node.right = del(node.right, e);
            return node;
        } else { // e.compareTo(node.e) == 0
            // 左子树为空的情况
            if (node.left == null) {
                Node tmpNode = node.right;
                node.right = null;
                size--;
                return tmpNode;
            }
            // 右子树为空的情况
            if (node.right == null) {
                Node tmpNode = node.left;
                node.left = null;
                size--;
                return tmpNode;
            }
            // 左右子树都不为空的情况
            // 找到最右子树最小的节点来补充到要删除的位置
            Node successor = min(node.right);
            successor.right = delMin(node.right);
            successor.left = node.left;

            node.right = node.left = null;

            return successor;
        }
    }

    @Override
    public String toString() {
        StringBuilder str = new StringBuilder();
        // str.append("null");
        
        generaTreeStr(root, 0, str);
        return str.toString();
    }
    private void generaTreeStr(Node node, int depth, StringBuilder str) {
        if (node == null) {
            str.append(generaString(depth)+"null\n");
            return;
        }
        str.append(generaString(depth) + node.e + "\n");
        generaTreeStr(node.left, depth++, str);
        generaTreeStr(node.right, depth++, str);
    }
    private String generaString(int depth) {
        StringBuilder str = new StringBuilder();
        for (int i=0; i<depth; i++) {
            str.append("--");
        }
        return str.toString();
    }


    public static void main(String[] argv) {
        BinarySearchTree<Integer> bst = new BinarySearchTree<>();
        int[] tree = {3,4,5,12,343,8,10,22};
        for (int i=0; i<tree.length; i++) {
            bst.add(tree[i]);
        }

        bst.preOrder();
        System.out.println();
        bst.preOrderNR();

       // System.out.println(bst);
    }
}